vignettes/customization.Rmd

```diff
@@ -0,0 +1,185 @@
+---
+title: "Customization"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Customization}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+```{r setup}
+library(pillar)
+```
+
+This draft describes a new backward-compatible design for making the appearance of a pillar more extensible.
+
+The design consists of two parts:
+
+1. Moving all formatting code into pillar
+1. Adding extension points for formatting individual pillars
+
+
+## Moving all formatting code into pillar
+
+### Advantages
+
+- Code that implements formatting for tables no longer needs tibble
+- All formatting code in one package
+- Simplify and document formatting code along the way
+- New `as_tbl()` for attaching the `"tbl"` class to arbitrary objects (e.g. data frames or data tables) for printing
+
+### Implementation
+
+`format.tbl()` retains its semantics.
+For compatibility it returns a character vector with one element per line.
+
+```{r}
+#' @export
+format.tbl <- function(x, ...) {
+  setup <- tbl_format_setup(x, width = width, n = n, n_extra = n_extra)
+  header <- tbl_format_header(x, setup)
+  body <- tbl_format_body(x, setup)
+  footer <- tbl_format_footer(x, setup)
+  c(header, body, footer)
+}
+```
+
+### New generics
+
+`tbl_format_setup()` takes the role of the (then superseded) `trunc_mat()`.
+The default mimics `trunc_mat()` but returns a stable data structure that is well documented.
+Classes that need to do more when setting up the printing implement a method.
+
+Contrary to the existing logic, the width would be baked into the `setup` object.
+I haven't seen a use case for formatting the same prepared object with multiple widths.
+
+```{r}
+tbl_format_setup.default <- function(x, ...) {
+  trunc_mat(x, ...)
+}
+```
+
+`tbl_format_header()` in the default implementation formats the output of `tbl_sum()`.
+The implementer is expected to apply `style_subtle()` in case the class wants to add color or other styling.
+
+```{r}
+tbl_format_header.default <- function(x, ...) {
+  sum <- tbl_sum(x)
+  style_subtle(
+    paste0(names(sum), ": ", sum)
+  )
+}
+```
+
+`tbl_format_body()` formats the colonnade prepared in the setup.
+The implementation is shown further below.
+
+`tbl_format_footer()` is copied from tibble.
+The implementer is expected to apply `style_subtle()` in case the class wants to add color or other styling.
+
+```{r}
+tbl_format_footer.default <- function(x, ...) {
+  style_subtle(format_footer(x, ...))
```

Comment:
Why `format_footer()`? That seems inconsistent with `header` and `body` methods?